### PR TITLE
Allow overriding when renaming

### PR DESCRIPTION
**Pull request by tomasmatus**
fixes: #468

Renaming file to a filename which already exists now causes the original file to be replaced with the renamed file.

For now this is only allowed for normal files and links to normal files since directories would require recursive merging.

image

### CODE REVIEW

**Comment by jelly** [comment]
> That's not how the inline confirmation should work. This makes it too easy to accidentally click a second time by having a second click on the rename and it becomes replace. It needs to be a secondary action that's in another spot to prevent accidental misclicks.
> 
> Here's how that concept works:
> 
> image
> 
> And conflicts with files and directories should be handled differently, as you cannot (nor should you) overwrite a directory with a file.

We already don't allow that, so then this should not offer that option. (Needs a good integration test)
 
> We also immediately know when files conflict, right? Someone doesn't have to click "Rename" first, right?

We do when loading the dialog.

**Comment by garrett** [comment]
>> We also immediately know when files conflict, right? Someone doesn't have to click "Rename" first, right?
>
> We do when loading the dialog.

We probably also want to handle the change, then change back to original name on the fly then too, like this new on on the bottom-left:

added change-then-change-back modal on bottom left

(To be clear, a modal should never show an error by default. This would only happen when someone changes the filename to something else and changes it back _or_ they submit the modal immediately with enter or click.)

File: src/dialogs/rename.tsx [hunk 1]
```diff
@@ -107,11 +137,12 @@ const RenameItemModal = ({ dialogResult, path, selected } : {
...
                 />}
                 <Form isHorizontal>
-                    <FormGroup fieldId="rename-item-input" label={_("New name")}>
+                    <FormGroup fieldId="rename-item-input" label={_("New filename")}>
                         <TextInput
                           value={name}
                           onChange={(_, val) => {
-                              setNameError(check_name(val, cwdInfo?.entries || {}));
+                              setNameError(checkName(val, cwdInfo?.entries || {}, selected));
+                              setOverrideFileName(checkCanOverride(val, cwdInfo?.entries || {}, selected));
```
Comment (cockpituous) [inline src/dialogs/rename.tsx:145]: These 2 added lines are not executed by any test.

File: src/dialogs/rename.tsx [hunk 2]
```diff
@@ -57,47 +76,58 @@ const RenameItemModal = ({ dialogResult, path, selected } : {
...
-                      { superuser: "try", err: "message" })
+        cockpit.spawn(mvCmd, { superuser: "try", err: "message" })
                 .then(() => {
                     dialogResult.resolve();
                 }, err => setErrorMessage(err.message));
     };
 
+    const title = (
+        <>
+            {_("Rename") + " "}
```
Comment (jelly) [inline src/dialogs/rename.tsx:97]: The design shows a question mark `Rename <b>Foo</b>?` here, I would suggest using `fmt_to_fragments`.

File: src/dialogs/rename.tsx (hunk 1 above)
Comment (tomasmatus) [reply src/dialogs/rename.tsx:145]: what

File: src/dialogs/rename.tsx [hunk 3]
```diff
@@ -107,11 +131,12 @@ const RenameItemModal = ({ dialogResult, path, selected } : {
...
                 />}
                 <Form isHorizontal>
-                    <FormGroup fieldId="rename-item-input" label={_("New name")}>
+                    <FormGroup fieldId="rename-item-input" label={_("New filename")}>
                         <TextInput
                           value={name}
                           onChange={(_, val) => {
-                              setNameError(check_name(val, cwdInfo?.entries || {}));
+                              setNameError(checkName(val, cwdInfo?.entries || {}, selected));
+                              setOverrideFileName(checkCanOverride(val, cwdInfo?.entries || {}, selected));
```
Comment (cockpituous) [inline src/dialogs/rename.tsx:139]: These 2 added lines are not executed by any test.

**Comment by jelly** [comment]
image

This feels a bit silly, but its a rather edge case we can do in a follow up.

**Comment by tomasmatus** [comment]
> lib is a symlink to a directory so this should not be allowed to be renamed.

huh yeah this needs to be fixed, I'll also make the tests more robust for all cases I can think of

**Review by jelly — CHANGES_REQUESTED** [review]

File: src/dialogs/rename.tsx [hunk 4]
```diff
@@ -29,25 +29,45 @@ import cockpit from 'cockpit';
...
+        }
+        return _("File exists");
     } else {
         return null;
     }
 }
 
+function checkCanOverride(candidate: string, entries: Record<string, FileInfo>, selectedFile: FolderFileInfo) {
+    if (candidate in entries) {
+        if (entries[candidate].type === "dir") {
```
[the remark this answers is not in the document]
Comment (tomasmatus) [reply src/dialogs/rename.tsx:60]: I am currently reworking the `checkCanOverride` function to fix checks for symlinks and I came to this too. The problem is that `FileInfo` has to `.to` prop which I kinda need

[the remark this answers is not in the document]
Comment (jelly) [reply src/dialogs/rename.tsx:60]: You mean FolderFileInfo has it, whole FileInfo does not have it?

[the remark this answers is not in the document]
Comment (tomasmatus) [reply src/dialogs/rename.tsx:60]: yes, FolderFileInfo extends FileInfo and adds `.to`

File: src/dialogs/rename.tsx [hunk 5]
```diff
@@ -107,11 +137,12 @@ const RenameItemModal = ({ dialogResult, path, selected } : {
...
                 />}
                 <Form isHorizontal>
-                    <FormGroup fieldId="rename-item-input" label={_("New name")}>
+                    <FormGroup fieldId="rename-item-input" label={label}>
                         <TextInput
                           value={name}
                           onChange={(_, val) => {
-                              setNameError(check_name(val, cwdInfo?.entries || {}));
+                              setNameError(checkName(val, cwdInfo?.entries || {}, selected));
+                              setOverrideFileName(checkCanOverride(val, cwdInfo?.entries || {}, selected));
```
Comment (cockpituous) [inline src/dialogs/rename.tsx:145]: These 2 added lines are not executed by any test.

File: test/check-application [hunk 6]
```diff
@@ -843,7 +844,7 @@ class TestFiles(testlib.MachineCase):
         b.wait_not_present("button.pf-m-danger")
         b.click("div.pf-v5-c-modal-box__close > button")
 
-        # Overwrite symlink to file
+        # Overwriting symlinks shows error
```
Comment (jelly) [inline test/check-application:847]: "Overwriting symlinks is not supported" sounds more technically true :)

**Review by jelly — CHANGES_REQUESTED** [review]
This looks good, I would combine both commits into one and update the pixel tests :+1:

File: src/dialogs/rename.tsx (hunk 5 above)
Comment (cockpituous) [inline src/dialogs/rename.tsx:145]: These 2 added lines are not executed by any test.

**Comment by tomasmatus** [comment]
rebased + squashed commits + updated pixel tests :+1:

**Review by garrett — CHANGES_REQUESTED** [review]
The UI looks good! Thanks!

There are some major interaction issues still:

1. When you select rename, the focus isn't there. You have to navigate to the field.
2. When you are in the input, hitting the enter key doesn't rename. Sometimes it seems to reload the page, other times it does nothing at all. It feels like the form isn't hooked up properly?

**Review by garrett — APPROVED** [review]
Overwrite, as added in this PR is fine. Thanks!

I opened a new issue with critical problems with the rename modal that should be worked on _immediately_ as a follow-up, based on my comments above.

**Review by jelly — APPROVED** [review]
Thanks!
